File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/TheOnlySelectionMatchedCategorizer.java
package cz.fb.manaus.reactor.categorizer;

import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Set;

@Component
public class TheOnlySelectionMatchedCategorizer implements SettledBetCategorizer {

    @Autowired
    private SelectionActualMatchedCategorizer selectionActualMatchedCategorizer;
    @Autowired
    private ActualMatchedCategorizer actualMatchedCategorizer;

    @Override
    public boolean isMarketSnapshotRequired() {
        return true;
    }

    @Override
    public Set<String> getCategories(SettledBet bet, BetCoverage coverage) {
        var selectionMatched = selectionActualMatchedCategorizer.getAmount(bet);
        var allMatched = actualMatchedCategorizer.getAmount(bet);
        if (selectionMatched.isPresent() && allMatched.isPresent()) {
            var theOnlyMatched = Price.amountEq(allMatched.getAsDouble(), selectionMatched.getAsDouble());
            return Set.of("theOnlyMatched_" + theOnlyMatched);
        } else {
            return Set.of();
        }
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/TheOnlySelectionMatchedCategorizer.kt
package cz.fb.manaus.reactor.categorizer

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.RealizedBet
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class TheOnlySelectionMatchedCategorizer : RealizedBetCategorizer {

    @Autowired
    private lateinit var selectionActualMatchedCategorizer: SelectionActualMatchedCategorizer
    @Autowired
    private lateinit var actualMatchedCategorizer: ActualMatchedCategorizer

    override val isMarketSnapshotRequired: Boolean = true

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val selectionMatched = selectionActualMatchedCategorizer.getAmount(realizedBet)
        val allMatched = actualMatchedCategorizer.getAmount(realizedBet)
        return if (selectionMatched != null && allMatched != null) {
            val theOnlyMatched = Price.amountEq(allMatched, selectionMatched)
            setOf("theOnlyMatched_$theOnlyMatched")
        } else {
            emptySet()
        }
    }
}
